File: src/test/java/leets/weeth/domain/comment/application/usecase/NoticeCommentUsecaseImplTest.java
package leets.weeth.domain.comment.application.usecase;

import leets.weeth.domain.board.domain.entity.Notice;
import leets.weeth.domain.board.domain.service.NoticeFindService;
import leets.weeth.domain.board.test.fixture.NoticeTestFixture;
import leets.weeth.domain.comment.application.dto.CommentDTO;
import leets.weeth.domain.comment.application.mapper.CommentMapper;
import leets.weeth.domain.comment.domain.entity.Comment;
import leets.weeth.domain.comment.domain.service.CommentFindService;
import leets.weeth.domain.comment.domain.service.CommentSaveService;
import leets.weeth.domain.comment.test.fixture.CommentTestFixture;
import leets.weeth.domain.file.application.mapper.FileMapper;
import leets.weeth.domain.file.domain.service.FileSaveService;
import leets.weeth.domain.user.application.exception.UserNotMatchException;
import leets.weeth.domain.user.domain.entity.User;
import leets.weeth.domain.user.domain.service.UserGetService;
import leets.weeth.domain.user.test.fixture.UserTestFixture;
import org.junit.jupiter.api.DisplayName;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.extension.ExtendWith;
import org.mockito.InjectMocks;
import org.mockito.Mock;
import org.mockito.junit.jupiter.MockitoExtension;

import java.util.List;

import static org.junit.jupiter.api.Assertions.assertThrows;
import static org.mockito.BDDMockito.*;
import static org.assertj.core.api.AssertionsForInterfaceTypes.assertThat;

@ExtendWith(MockitoExtension.class)
class NoticeCommentUsecaseImplTest {
    @InjectMocks NoticeCommentUsecaseImpl noticeCommentUsecase;

    @Mock CommentSaveService commentSaveService;
    @Mock CommentFindService commentFindService;

    @Mock FileSaveService fileSaveService;
    @Mock FileMapper fileMapper;

    @Mock NoticeFindService noticeFindService;

    @Mock UserGetService userGetService;
    @Mock CommentMapper commentMapper;

    @Test
    @DisplayName("부모 댓글이 없는 공지사항 댓글 작성")
    void saveNoticeComment() {
        Long userId = 1L;
        Long noticeId = 1L;
        Long commentId = 1L;
        // given
        User user  = UserTestFixture.createActiveUser1(1L);
        Notice notice = NoticeTestFixture.createNotice(noticeId, "제목1");

        CommentDTO.Save dto = new CommentDTO.Save(null, "댓글1", List.of());

        Comment comment = CommentTestFixture.createComment(commentId, dto.content(), user, notice);

        given(commentMapper.fromCommentDto(dto, notice, user, null)).willReturn(comment);
        given(userGetService.find(user.getId())).willReturn(user);
        given(noticeFindService.find(notice.getId())).willReturn(notice);
        given(fileMapper.toFileList(dto.files(), comment)).willReturn(List.of());

        // when
        noticeCommentUsecase.saveNoticeComment(dto, noticeId, userId);

        // then
        verify(userGetService).find(userId);
        verify(noticeFindService).find(noticeId);
        verify(commentSaveService).save(comment);

        assertThat(notice.getComments()).contains(comment);

    }

    @Test
    @DisplayName("부모 댓글이 있는 경우 공지사항 댓글 작성")
    void saveChildNoticeComment() {
        Long userId = 1L;
        Long noticeId = 1L;
        Long parentCommentId = 1L;
        Long childCommentId = 2L;
        // given
        User user  = UserTestFixture.createActiveUser1(parentCommentId);
        Notice notice = NoticeTestFixture.createNotice(noticeId, "제목1");

        Comment parentComment = CommentTestFixture.createComment(parentCommentId, "부모 댓글", user, notice);

        CommentDTO.Save childCommentDTO = new CommentDTO.Save(parentCommentId, "자식 댓글", List.of());
        Comment childComment = CommentTestFixture.createComment(childCommentId, childCommentDTO.content(), user, notice);

        given(commentMapper.fromCommentDto(childCommentDTO, notice, user, parentComment)).willReturn(childComment);
        given(userGetService.find(user.getId())).willReturn(user);
        given(commentFindService.find(parentComment.getId())).willReturn(parentComment);
        given(noticeFindService.find(notice.getId())).willReturn(notice);
        given(fileMapper.toFileList(childCommentDTO.files(), childComment)).willReturn(List.of());

        // when
        noticeCommentUsecase.saveNoticeComment(childCommentDTO, noticeId, userId);

        // then
        verify(commentFindService).find(parentComment.getId());
        verify(commentSaveService).save(childComment);

        assertThat(parentComment.getChildren()).contains(childComment);
    }

    @Test
    @DisplayName("공지사항 댓글 수정 시 작성자와 수정 요청자가 다르면 예외가 발생한다")
    void updateNoticeComment_throwsException_whenUserIsNotOwner() {
        Long different = 2L;
        Long noticeId = 1L;
        Long commentId = 1L;
        // given
        User user  = UserTestFixture.createActiveUser1(1L);
        User user2  = UserTestFixture.createActiveUser1(2L);
        Notice notice = NoticeTestFixture.createNotice(noticeId, "제목1");

        CommentDTO.Update dto = new CommentDTO.Update("수정 완료", List.of());

        Comment comment = CommentTestFixture.createComment(commentId, dto.content(), user, notice);

        given(userGetService.find(user2.getId())).willReturn(user2);
        given(noticeFindService.find(notice.getId())).willReturn(notice);
        given(commentFindService.find(comment.getId())).willReturn(comment);

        // when & then
        assertThrows(UserNotMatchException.class, () ->
                noticeCommentUsecase.updateNoticeComment(dto, noticeId, comment.getId(), different)
        );
    }
}

File: src/test/kotlin/com/weeth/domain/comment/application/usecase/NoticeCommentUsecaseImplTest.kt
package com.weeth.domain.comment.application.usecase

import com.weeth.domain.board.domain.service.NoticeFindService
import com.weeth.domain.board.fixture.NoticeTestFixture
import com.weeth.domain.comment.application.dto.CommentDTO
import com.weeth.domain.comment.application.mapper.CommentMapper
import com.weeth.domain.comment.domain.service.CommentDeleteService
import com.weeth.domain.comment.domain.service.CommentFindService
import com.weeth.domain.comment.domain.service.CommentSaveService
import com.weeth.domain.comment.fixture.CommentTestFixture
import com.weeth.domain.file.application.mapper.FileMapper
import com.weeth.domain.file.domain.service.FileDeleteService
import com.weeth.domain.file.domain.service.FileGetService
import com.weeth.domain.file.domain.service.FileSaveService
import com.weeth.domain.user.application.exception.UserNotMatchException
import com.weeth.domain.user.domain.service.UserGetService
import com.weeth.domain.user.fixture.UserTestFixture
import io.kotest.assertions.throwables.shouldThrow
import io.kotest.core.spec.style.DescribeSpec
import io.kotest.matchers.collections.shouldContain
import io.mockk.every
import io.mockk.mockk
import io.mockk.verify

class NoticeCommentUsecaseImplTest :
    DescribeSpec({

        val commentSaveService = mockk<CommentSaveService>(relaxUnitFun = true)
        val commentFindService = mockk<CommentFindService>()
        val commentDeleteService = mockk<CommentDeleteService>(relaxUnitFun = true)
        val fileSaveService = mockk<FileSaveService>(relaxUnitFun = true)
        val fileGetService = mockk<FileGetService>()
        val fileDeleteService = mockk<FileDeleteService>(relaxUnitFun = true)
        val fileMapper = mockk<FileMapper>()
        val noticeFindService = mockk<NoticeFindService>()
        val userGetService = mockk<UserGetService>()
        val commentMapper = mockk<CommentMapper>()

        val noticeCommentUsecase =
            NoticeCommentUsecaseImpl(
                commentSaveService,
                commentFindService,
                commentDeleteService,
                fileSaveService,
                fileGetService,
                fileDeleteService,
                fileMapper,
                noticeFindService,
                userGetService,
                commentMapper,
            )

        describe("saveNoticeComment") {
            it("부모 댓글이 없는 공지사항 댓글 작성") {
                val userId = 1L
                val noticeId = 1L
                val commentId = 1L

                val user = UserTestFixture.createActiveUser1(1L)
                val notice = NoticeTestFixture.createNotice(id = noticeId, title = "제목1")

                val dto = CommentDTO.Save(null, "댓글1", listOf())

                val comment = CommentTestFixture.createComment(commentId, dto.content(), user, notice)

                every { commentMapper.fromCommentDto(dto, notice, user, null) } returns comment
                every { userGetService.find(user.id) } returns user
                every { noticeFindService.find(notice.id) } returns notice
                every { fileMapper.toFileList(dto.files(), comment) } returns listOf()

                noticeCommentUsecase.saveNoticeComment(dto, noticeId, userId)

                verify { userGetService.find(userId) }
                verify { noticeFindService.find(noticeId) }
                verify { commentSaveService.save(comment) }

                notice.comments shouldContain comment
            }

            it("부모 댓글이 있는 경우 공지사항 댓글 작성") {
                val userId = 1L
                val noticeId = 1L
                val parentCommentId = 1L
                val childCommentId = 2L

                val user = UserTestFixture.createActiveUser1(parentCommentId)
                val notice = NoticeTestFixture.createNotice(id = noticeId, title = "제목1")

                val parentComment = CommentTestFixture.createComment(parentCommentId, "부모 댓글", user, notice)

                val childCommentDTO = CommentDTO.Save(parentCommentId, "자식 댓글", listOf())
                val childComment = CommentTestFixture.createComment(childCommentId, childCommentDTO.content(), user, notice)

                every { commentMapper.fromCommentDto(childCommentDTO, notice, user, parentComment) } returns childComment
                every { userGetService.find(user.id) } returns user
                every { commentFindService.find(parentComment.id) } returns parentComment
                every { noticeFindService.find(notice.id) } returns notice
                every { fileMapper.toFileList(childCommentDTO.files(), childComment) } returns listOf()

                noticeCommentUsecase.saveNoticeComment(childCommentDTO, noticeId, userId)

                verify { commentFindService.find(parentComment.id) }
                verify { commentSaveService.save(childComment) }

                parentComment.children shouldContain childComment
            }
        }

        describe("updateNoticeComment") {
            it("공지사항 댓글 수정 시 작성자와 수정 요청자가 다르면 예외가 발생한다") {
                val different = 2L
                val noticeId = 1L
                val commentId = 1L

                val user = UserTestFixture.createActiveUser1(1L)
                val user2 = UserTestFixture.createActiveUser1(2L)
                val notice = NoticeTestFixture.createNotice(id = noticeId, title = "제목1")

                val dto = CommentDTO.Update("수정 완료", listOf())
                val comment = CommentTestFixture.createComment(commentId, dto.content(), user, notice)

                every { userGetService.find(user2.id) } returns user2
                every { noticeFindService.find(notice.id) } returns notice
                every { commentFindService.find(comment.id) } returns comment

                shouldThrow<UserNotMatchException> {
                    noticeCommentUsecase.updateNoticeComment(dto, noticeId, comment.id, different)
                }
            }
        }
    })
